NWNXLib/Services/Events/Events.cpp

```diff
@@ -149,23 +157,21 @@ void EventsProxy::ClearEvent(const std::string& eventName)
     m_proxyBase.ClearEvent(std::move(concreteToken));
 }
 
-
-
-template<> std::optional<int32_t>&              Events::Argument::Get<int32_t>()             { return m_int; }
-template<> std::optional<float>&                Events::Argument::Get<float>()               { return m_float; }
-template<> std::optional<API::Types::ObjectID>& Events::Argument::Get<API::Types::ObjectID>(){ return m_object; }
-template<> std::optional<std::string>&          Events::Argument::Get<std::string>()         { return m_string; }
-template<> std::optional<CGameEffect*>&         Events::Argument::Get<CGameEffect*>()        { return m_effect; }
+namespace detail
+{
+template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+} // namespace detail
 
 std::string Events::Argument::toString() const
 {
-    if (m_int)    return std::to_string(*m_int);
-    if (m_float)  return std::to_string(*m_float);
-    if (m_object) return Utils::ObjectIDToString(*m_object);
-    if (m_string) return *m_string;
-    if (m_effect) return *m_effect ? std::string("EffectID:") + std::to_string((*m_effect)->m_nID) : std::string("nullptr effect");
-
-    return std::string("");
+    return std::visit(detail::overloaded {
+        [](auto arg) { return std::to_string(arg); },
+        [](API::Types::ObjectID arg) { return Utils::ObjectIDToString(arg); },
+        [](const std::string& arg) { return arg; },
+        [](NullArgument) { return std::string("(null)"); },
+        [](CGameEffect* arg) { return arg ? std::string("EffectID:") + std::to_string(arg->m_nID) : std::string("nullptr effect");}
+    }, m_data);
```

Comment:
This seems like a step backwords to me. It's about the same verbosity, but much less clearer. It raises the C++ knowledge bar to follow this code quite a lot (I personally know dozens of people who would understand the old version but not the new). And I can't even imagine what sort of eldritch errors it will spew out if something goes wrong.